### DISCUSSION TITLE
Multiband nodata masks

### DISCUSSION THREAD

**Discussion by BielStela** [cogeotiff/rio-tiler#570] [General]
Hi! 
I found a problem with how the nodata mask is handled for multiband rasters. It happens that I have a multiband raster with no internal mask, only a global nodata value, and I expect that when asking for a tile of a specific band, the mask is the one for that band. Instead  what I get is the mask for the first band only. Looking at the code I see that deep in the low level `read` function uses 
```python
data = dataset.read(
    indexes=indexes,
    window=window,
    out_shape=(len(indexes), height, width) if height and width else None,
    resampling=resampling,
    boundless=boundless,
)
mask = dataset.dataset_mask(
    window=window,
    out_shape=(height, width) if height and width else None,
    resampling=resampling,
    boundless=boundless,
)

```
Is there any major reason on why multiband rasters can have only one mask representation for all bands?

I'm sure it is not as simple as to use `dataset.read_masks` with the `indexes` param because it will return a mask with an extra axis and that will break lots of things down the line. But maybe is it something worth investigating. 

Thank you!

**Comment by vincentsarago** [comment]
@BielStela see https://github.com/rasterio/rasterio/blob/main/rasterio/_io.pyx#L991-L1010 for how the mask is created from multi band dataset with internal nodata value. 

>  Instead what I get is the mask for the first band only.

are you sure, to me it seem the mask should be the combination (https://github.com/rasterio/rasterio/blob/main/rasterio/_io.pyx#L1035) of all band's mask 🤷‍♂️
